Mathlib/GroupTheory/GroupAction/Defs.lean

```diff
@@ -0,0 +1,1194 @@
+    (hf : Surjective f) (smul : ∀ (c : M) (x), f (c • x) = c • f x) : DistribMulAction M B :=
+  { hf.distribSmul f smul, hf.mulAction f smul with smul := (· • ·) }
+#align function.surjective.distrib_mul_action Function.Surjective.distribMulAction
+
+/-- Push forward the action of `R` on `M` along a compatible surjective map `f : R →* S`.
+
+See also `function.surjective.mul_action_left` and `function.surjective.module_left`.
+-/
+@[reducible]
+def Function.Surjective.distribMulActionLeft {R S M : Type _} [Monoid R] [AddMonoid M]
+    [DistribMulAction R M] [Monoid S] [SMul S M] (f : R →* S) (hf : Function.Surjective f)
+    (hsmul : ∀ (c) (x : M), f c • x = c • x) : DistribMulAction S M :=
+  { hf.distribSmulLeft f hsmul, hf.mulActionLeft f hsmul with smul := (· • ·) }
+#align function.surjective.distrib_mul_action_left Function.Surjective.distribMulActionLeft
+
+variable (A)
+
+/-- Compose a `distrib_mul_action` with a `monoid_hom`, with action `f r' • m`.
+See note [reducible non-instances]. -/
+@[reducible]
+def DistribMulAction.compHom [Monoid N] (f : N →* M) : DistribMulAction N A :=
+  { DistribSmul.compFun A f, MulAction.compHom A f with smul := SMul.comp.smul f }
+#align distrib_mul_action.comp_hom DistribMulAction.compHom
+
+/-- Each element of the monoid defines a additive monoid homomorphism. -/
+@[simps]
+def DistribMulAction.toAddMonoidHom (x : M) : A →+ A :=
+  DistribSmul.toAddMonoidHom A x
+#align distrib_mul_action.to_add_monoid_hom DistribMulAction.toAddMonoidHom
+
+variable (M)
+
+/-- Each element of the monoid defines an additive monoid homomorphism. -/
+@[simps]
+def DistribMulAction.toAddMonoidEnd :
+    M →* AddMonoid.End A where
+  toFun := DistribMulAction.toAddMonoidHom A
+  map_one' := AddMonoidHom.ext <| one_smul M
+  map_mul' x y := AddMonoidHom.ext <| mul_smul x y
+#align distrib_mul_action.to_add_monoid_End DistribMulAction.toAddMonoidEnd
+
+instance AddMonoid.nat_smulCommClass :
+    SMulCommClass ℕ M
+      A where smul_comm n x y := ((DistribMulAction.toAddMonoidHom A x).map_nsmul y n).symm
+#align add_monoid.nat_smul_comm_class AddMonoid.nat_smulCommClass
+
+-- `SMulCommClass.symm` is not registered as an instance, as it would cause a loop
+instance AddMonoid.nat_smul_comm_class' : SMulCommClass M ℕ A :=
+  SMulCommClass.symm _ _ _
+#align add_monoid.nat_smul_comm_class' AddMonoid.nat_smul_comm_class'
+
+end
+
+section
+
+variable [Monoid M] [AddGroup A] [DistribMulAction M A]
+
+instance AddGroup.int_smul_comm_class :
+    SMulCommClass ℤ M
+      A where smul_comm n x y := ((DistribMulAction.toAddMonoidHom A x).map_zsmul y n).symm
+#align add_group.int_smul_comm_class AddGroup.int_smul_comm_class
+
+-- `SMulCommClass.symm` is not registered as an instance, as it would cause a loop
+instance AddGroup.int_smul_comm_class' : SMulCommClass M ℤ A :=
+  SMulCommClass.symm _ _ _
+#align add_group.int_smul_comm_class' AddGroup.int_smul_comm_class'
```

Comment:
```suggestion
instance AddGroup.int_smulCommClass :
    SMulCommClass ℤ M
      A where smul_comm n x y := ((DistribMulAction.toAddMonoidHom A x).map_zsmul y n).symm
#align add_group.int_smul_comm_class AddGroup.int_smulCommClass

-- `SMulCommClass.symm` is not registered as an instance, as it would cause a loop
instance AddGroup.int_smulCommClass' : SMulCommClass M ℤ A :=
  SMulCommClass.symm _ _ _
#align add_group.int_smul_comm_class' AddGroup.int_smulCommClass'
```